File: Section06OptionalChallenges/DivideAndConquer_countZeros_v2.js
function countZeroes(arr) {
  let firstZero = findFirstNum(arr, 0);
  return firstZero === -1 ? 0 : arr.length - firstZero;
}

function findFirstNum(arr, num, lowIndex = 0, highIndex = arr.length - 1) {
  if (highIndex >= lowIndex) {
    let midIndex = lowIndex + Math.floor((highIndex - lowIndex) / 2);
    if (arr[midIndex] === num && (midIndex === 0 || arr[midIndex - 1] === 1)) {
      return midIndex;
    } else if (arr[midIndex] === 1) {
      return findFirstNum(arr, num, midIndex + 1, highIndex);
    } else {
      return findFirstNum(arr, num, lowIndex, midIndex - 1);
    }
  }
  return -1;
}

console.log(countZeroes([1,1,1,1,0,0]));
console.log(countZeroes([1,0,0,0,0]));
console.log(countZeroes([0,0,0]));
console.log(countZeroes([1,1,1,1]));

File: Section06OptionalChallenges/Section06OptionalChallenges_DOTNET/DivideAndConquer_CountZeros_v2.cs
using System;

namespace Section06OptionalChallenges.DivideAndConquer;

public partial class DivideAndConquer
{
  public static int CountZerosRecursive(int[] arr)
  {
    var firstZero = FindFirstNum(arr, 0, 0, arr.Length - 1);
    return firstZero == -1 ? 0 : arr.Length - firstZero;
  }

  private static int FindFirstNum(int[] arr, int num, int lowIndex, int highIndex)
  {
    if (highIndex >= lowIndex)
    {
      var midIndex = lowIndex + (int)Math.Floor((double)(highIndex - lowIndex) / 2);
      if (arr[midIndex] == 0 && (midIndex == 0 || arr[midIndex - 1] == 1))
      {
        return midIndex;
      }
      else if (arr[midIndex] == 1)
      {
        return FindFirstNum(arr, num, midIndex + 1, highIndex);
      }
      else
      {
        return FindFirstNum(arr, num, lowIndex, midIndex - 1);
      }
    }
    return -1;
  }
}